File: tools/node/validate/validate.js
#!/usr/bin/env node
/**
 * Generic JSON Schema validator used by the specialized validators:
 *
 * - input:  AOM input surfaces  (spec/v0.1.0/aom-input-schema.json)
 * - output: AOM outputs         (spec/v0.1.0/aom-output-schema.json)
 * - site:   Site policy         (spec/v0.1.0/site-policy-schema.json)
 *
 * Usage (internal; prefer the three wrappers):
 *   node validate.js <schema-path> <json-path>
 */
import Ajv from "ajv/dist/2020.js";
import addFormats from "ajv-formats";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, "..", "..", "..");

function loadJson(p) {
  try {
    return JSON.parse(fs.readFileSync(p, "utf-8"));
  } catch (e) {
    console.error(`[FAIL] JSON parse error in ${p}: ${e.message}`);
    process.exit(1);
  }
}

const args = process.argv.slice(2);
if (!args[0] || !args[1]) {
  console.error("Usage: node validate.js <schema-path> <json-path>");
  process.exit(1);
}

const schemaPath = path.resolve(args[0]);
const jsonPath = path.resolve(args[1]);

if (!fs.existsSync(schemaPath) || !fs.statSync(schemaPath).isFile()) {
  console.error("Schema not found: " + schemaPath);
  process.exit(1);
}
if (!fs.existsSync(jsonPath) || !fs.statSync(jsonPath).isFile()) {
  console.error("JSON not found: " + jsonPath);
  process.exit(1);
}

const schema = loadJson(schemaPath);
const data = loadJson(jsonPath);
const ajv = new Ajv({ strict: false, allErrors: true });
addFormats(ajv);
const validate = ajv.compile(schema);

if (validate(data)) {
  console.log("[OK] " + jsonPath + " is valid.");
  process.exit(0);
}
console.error("[FAIL] " + jsonPath + " is NOT valid:");

function getNodeAtPath(root, instancePath) {
  if (!instancePath) return root;
  const parts = instancePath.split("/").slice(1); // skip leading empty segment
  let node = root;
  for (const part of parts) {
    if (node == null) return undefined;
    const key = Number.isInteger(Number(part)) && part !== "" ? Number(part) : part;
    node = node[key];
  }
  return node;
}

(validate.errors || [])
  .sort((a, b) => (a.instancePath || "").localeCompare(b.instancePath || ""))
  .forEach((e) => {
    const pathStr = e.instancePath || "<root>";
    console.error("- " + pathStr + ": " + e.message);
    const node = getNodeAtPath(data, e.instancePath || "");
    if (node !== undefined) {
      let snippet;
      try {
        snippet = JSON.stringify(node, null, 2);
      } catch {
        snippet = String(node);
      }
      if (snippet.length > 400) snippet = snippet.slice(0, 397) + "...";
      console.error("  context:");
      snippet.split("\n").forEach((line) => console.error("    " + line));
    }
  });
process.exit(1);

File: tools/python/validate/validate.py
#!/usr/bin/env python3
"""
Generic JSON Schema validator used by the specialized validators:

- input:  AOM input surfaces  (spec/v0.1.0/aom-input-schema.json)
- output: AOM outputs         (spec/v0.1.0/aom-output-schema.json)
- site:   Site policy         (spec/v0.1.0/site-policy-schema.json)

Usage (internal; prefer the three wrappers):
  python validate.py <schema-path> <json-path>
"""

import json
import sys
from pathlib import Path

from jsonschema import Draft202012Validator

_SCRIPT_DIR = Path(__file__).resolve().parent
REPO_ROOT = _SCRIPT_DIR.parent.parent.parent.parent


def load_json(path: Path):
    try:
        with path.open("r", encoding="utf-8-sig") as f:
            return json.load(f)
    except json.JSONDecodeError as e:
        # Make it very clear which file failed to parse.
        print(f"[FAIL] JSON parse error in {path}: {e}")
        sys.exit(1)


def main():
    args = sys.argv[1:]
    if len(args) != 2:
        print("Usage: python validate.py <schema-path> <json-path>")
        sys.exit(1)

    schema_path = Path(args[0]).resolve()
    json_path = Path(args[1]).resolve()

    if not schema_path.is_file():
        print(f"Schema file not found: {schema_path}")
        sys.exit(1)

    if not json_path.is_file():
        print(f"JSON file not found: {json_path}")
        sys.exit(1)

    schema = load_json(schema_path)
    data = load_json(json_path)

    validator = Draft202012Validator(schema)
    errors = sorted(validator.iter_errors(data), key=lambda e: e.path)

    if not errors:
        print(f"[OK] {json_path} is valid.")
        sys.exit(0)

    print(f"[FAIL] {json_path} is NOT valid:")
    for err in errors:
        path_list = list(err.path)
        path_str = ".".join(str(p) for p in path_list) or "<root>"
        print(f"- {path_str}: {err.message}")

        # Best-effort: show a small JSON snippet at the error location to help humans.
        try:
            node = data
            for p in path_list:
                node = node[p]
        except Exception:
            node = None

        if node is not None:
            try:
                snippet = json.dumps(node, indent=2, ensure_ascii=False)
            except TypeError:
                snippet = repr(node)
            if len(snippet) > 400:
                snippet = snippet[:397] + "..."
            print("  context:")
            for line in snippet.splitlines():
                print(f"    {line}")

    sys.exit(1)


if __name__ == "__main__":
    main()
